### ISSUE TITLE
Check Pandas and Eli5 

### ISSUE THREAD

**Issue by nareshshah139** [TeamHG-Memex/eli5#167]
Hello kmike. Could you re-check this piece of code? and tell me possible reasons for the error?

**Comment by nareshshah139** [comment]
https://github.com/nareshshah139/titanic_rebranded/blob/master/ELI5_Example.ipynb

**Comment by nareshshah139** [comment]
Still think there is something wrong with eli5/xgboost/integration with pandas here.

**Comment by lopuhin** [comment]
@nareshshah139  I tried running the second part of the notebook, there is indeed an issue with pandas + eli5 + xgboost here (I didn't check yet if sklearn + eli5 + pandas work).

There are even two issues here I think: first is that we don't know how to turn a single item (``valid_xs[order_list].loc[1]``) into a dataframe that is suitable to be passed into xgboost (that should be ``valid_xs[order_list].loc[[1]]``). This is the reason for ``ValueError: feature_names mismatch`` error you mentioned:
```
/home/kostia/shub/memex/eli5/eli5/xgboost.py in explain_prediction_xgboost(xgb, doc, vec, top, top_targets, target_names, targets, feature_names, feature_re, feature_filter, vectorized)
    135         X = X.tocsc()
    136 
--> 137     proba = predict_proba(xgb, X)
    138     scores_weights = _prediction_feature_weights(xgb, X, feature_names)
    139 

/home/kostia/shub/memex/eli5/eli5/sklearn/utils.py in predict_proba(estimator, X)
     47     if is_probabilistic_classifier(estimator):
     48         try:
---> 49             proba, = estimator.predict_proba(X)
     50             return proba
     51         except NotImplementedError:

/home/kostia/shub/memex/eli5/venv/lib/python3.5/site-packages/xgboost/sklearn.py in predict_proba(self, data, output_margin, ntree_limit)
    475         class_probs = self.booster().predict(test_dmatrix,
    476                                              output_margin=output_margin,
--> 477                                              ntree_limit=ntree_limit)
    478         if self.objective == "multi:softprob":
    479             return class_probs

/home/kostia/shub/memex/eli5/venv/lib/python3.5/site-packages/xgboost/core.py in predict(self, data, output_margin, ntree_limit, pred_leaf)
    937             option_mask |= 0x02
    938 
--> 939         self._validate_features(data)
    940 
    941         length = ctypes.c_ulong()

/home/kostia/shub/memex/eli5/venv/lib/python3.5/site-packages/xgboost/core.py in _validate_features(self, data)
   1177 
   1178                 raise ValueError(msg.format(self.feature_names,
-> 1179                                             data.feature_names))
   1180 
   1181     def get_split_value_histogram(self, feature, fmap='', bins=None, as_pandas=True):

ValueError: feature_names mismatch
```

But even if I pass a whole dataframe with one row into explain prediction (``valid_xs[order_list].loc[[1]]``), another error happens:
```
/home/kostia/shub/memex/eli5/eli5/xgboost.py in explain_prediction_xgboost(xgb, doc, vec, top, top_targets, target_names, targets, feature_names, feature_re, feature_filter, vectorized)
    136 
    137     proba = predict_proba(xgb, X)
--> 138     scores_weights = _prediction_feature_weights(xgb, X, feature_names)
    139 
    140     x, = add_intercept(X)

/home/kostia/shub/memex/eli5/eli5/xgboost.py in _prediction_feature_weights(xgb, X, feature_names)
    193             ) for target_idx in range(n_targets)]
    194     else:
--> 195         scores_weights = [target_feature_weights(leaf_ids, tree_dumps)]
    196     return scores_weights
    197 

/home/kostia/shub/memex/eli5/eli5/xgboost.py in _target_feature_weights(leaf_ids, tree_dumps, feature_names)
    214         for node, child in zip(path, path[1:]):
    215             f_num_match = re.search('^f(\d+)$', node['split'])
--> 216             idx = int(f_num_match.groups()[0])
    217             feature_weights[idx] += child['leaf'] - node['leaf']
    218         # Root "leaf" value is interpreted as bias

AttributeError: 'NoneType' object has no attribute 'groups'
```
This is due to the fact that we expect default feature names, but a pandas dataframe provides custom feature names instead, and we can not parse them.

**Comment by lopuhin** [comment]
pandas is a popular data science library, I think it would be nice to support it

**Comment by lopuhin** [comment]
it would be also nice if we wouldn't need a vectorizer to get proper feature names with pandas, we would be able to pull them from the dataframe.

**Comment by nareshshah139** [comment]
Thanks! This works now.

**Comment by carsonyan** [comment]
I'm still having this problem when trying to pass the pandas series:

ValueError: DataFrame.dtypes for data must be int, float or bool.
Did not expect the data types in fields ExternalRiskEstimate, MSinceOldestTradeOpen, MSinceMostRecentTradeOpen, AverageMInFile, NumSatisfactoryTrades, NumTrades60Ever2DerogPubRec, NumTrades90Ever2DerogPubRec, PercentTradesNeverDelq, MSinceMostRecentDelq, NumTotalTrades, NumTradesOpeninLast12M, PercentInstallTrades, MSinceMostRecentInqexcl7days, NumInqLast6M, NumInqLast6Mexcl7days, NetFractionRevolvingBurden, NetFractionInstallBurden, NumRevolvingTradesWBalance, NumInstallTradesWBalance, NumBank2NatlTradesWHighUtilization, PercentTradesWBalance, MaxDelq2PublicRecLast12M_tm, MaxDelqEver_tm
